### ISSUE TITLE
Rule properties defined in xml ruleset not used by MethodNamingConventionsRule instance

### ISSUE THREAD

**Issue by jonasgeiregat** [pmd/pmd#1811]
**Affects PMD Version:**  6.14.0

**Rule:** MethodNamingConventionsRule

**Description:** 

When defining a custom junit4TestPattern it is not used by MethodNamingConventionsRule. I have attached the debugger process and found out that my defined property is present in the instance but I was unable to figure out why it's not used.
It is present in the propertyValuesByDescriptor instance variable.
I'm taking a wild guess here but instead of using the default junit4Regexp value for checking on a match shouldn't the propertyValueByDescriptor be checked before ?

For example

```
checkMatches(node, propertyValuesByDescriptor.getOrDefault(junit4Regex, junit4Regex)
, data);

// instead of

checkMatches(node, junit4Regex, data);
```

**Code Sample demonstrating the issue:**

<img width="2972" alt="Screen Shot 2019-05-07 at 16 31 22" src="https://user-images.githubusercontent.com/288105/57308176-23fbbc00-70e6-11e9-8de0-4c81ef5b7249.png">

**Running PMD through:** *[ Gradle ]*

**Comment by adangel** [comment]
Hm... on first glance, it looks correct.
Which message do you get exactly? Is it "The JUnit 4 test method name 'wrongTestMethod' doesn't match '[a-z]*_[A-Z]{1}[a-z,A-Z]*_[A-Z]{1}[a-z,A-Z]*'"?

junit4Regex is not the value, it is only the property descriptor. This is used in `checkMatches` to lookup the actual value.

The following test case passed:

```xml
    <test-code>
        <description>junit4 test</description>
        <rule-property name="junit4TestPattern">[a-z]*_[A-Z]{1}[a-z,A-Z]*_[A-Z]{1}[a-z,A-Z]*</rule-property>
        <expected-problems>11</expected-problems>
        <expected-linenumbers>6</expected-linenumbers>
        <code><![CDATA[
            import org.junit.Test;

            public class Foo {

                @Test
                public void wrongTestMethod() {
                }

                @Test
                public void test_FooTest_ShouldWork() {
                }
            }
            ]]>
        </code>
    </test-code>
```

So, I suspect something with auxclasspath. In order to detect junit4 methods, JUnit4 has to be on the auxiliary classpath.

Could you share your build.gradle and the offending code snippet?

**Comment by jonasgeiregat** [comment]
I've done some more debugging and I'm convinced it's not related to JUnit being missing on the auxiliary classpath. Because `this.isJunit4Test(node)` evaluates to true and `                this.checkMatches(node, this.junit4Regex, data);` is called. Stepping into the method and following the code path down I could see that the correct regex was used and `this.addViolation(data, node, ..`was called. But I'm unable to step into this method (for now) to investigate further.

So I think the question is why isn't the violation being added trigger a failing build ?

An extract of my build.gradle file:

`
    apply plugin: 'pmd'
    pmd {
        toolVersion = '6.14.0'
        ruleSetFiles = files(project.rootDir.toString() + "/etc/pmd/rules-set.xml")
        ruleSets = []
    }
`

**Comment by oowekyala** [comment]
@jonasgeiregat I don't think it's a bug, your violationSuppressXPath just returns true all the time. Remember, that it's evaluated with the violating node as the context node of the expression.

```ruby
//ClassOrInterfaceDeclaration
```

Since this starts with `//`, it will match any ClassOrInterfaceDeclaration in the same file as the violation, meaning, not necessarily one that encloses the violation node.

```ruby
['.*PrincipalRepository']
```

This predicate is always true, because any non-empty string is truthy (you should be testing an attribute of the node). A predicate like `['foo']` is evaluated by converting the string to a boolean with the `boolean` function, which yields `true()` if the string is non-empty.

So basically the predicate doesn't test anything. The effect, is that your `violationSuppressXPath` returns true (ie suppresses the violation) any time the compilation unit where the violation is found contains some ClassOrInterfaceDeclaration, which is pretty frequent.

To make this work, you could replace it with
```ruby
./ancestor::ClassOrInterfaceDeclaration[@Image = 'PrincipalRepository']
```

Note that unfortunately, XPath 1.0 does not support regular expressions so you can't do a regex test. In this instance, I think a `@SuppressWarnings("PMD.MethodNamingConventions")` like you came up with is more appropriate.
